lib/graphql/schema/warden.rb

```diff
@@ -126,8 +126,53 @@ def root_type_for_operation(op_name)
 
       private
 
+      def union_memberships(obj_type)
+        @unions ||= read_through { |obj_type| @schema.union_memberships(obj_type).select { |u| visible?(u) } }
+        @unions[obj_type]
+      end
+
       def visible_field?(field_defn)
-        visible?(field_defn) && visible?(field_defn.type.unwrap)
+        visible?(field_defn) && visible_type?(field_defn.type.unwrap)
+      end
+
+      def visible_type?(type_defn)
+        visible?(type_defn) &&
+        has_visible_fields?(type_defn) && (
```

Comment:
thought of this scenario ^ if all fields are hidden on a type, the type should be hidden 🙈

Reply:
This seems ripe for an infinite loop if a type returns a field of itself.

Reply:
good catch 🤔 🙉 

Maybe this should be cached